clinica/iotools/converters/ukb_to_bids/ukb_utils.py

```diff
@@ -0,0 +1,305 @@
+        .apply(pd.Series)
+    )
+    df_clinical = df_clinical.assign(
+        bids_full_path=lambda df: df.participant_id
+        + "/"
+        + df.session
+        + "/"
+        + df.datatype
+        + "/"
+        + df.bids_filename
+    )
+    df_clinical = df_clinical.assign(
+        bids_sidecar_path=lambda df: df.participant_id
+        + "/"
+        + df.session
+        + "/"
+        + df.datatype
+        + "/"
+        + df.sidecar_filename
+    )
+    return df_clinical
+
+
+def dataset_to_bids(df_clinical: DataFrame) -> Tuple[DataFrame, DataFrame, DataFrame]:
+
+    import os
+
+    import pandas as pd
+
+    # open the reference for building the tsvs:
+    path_to_ref_csv = os.path.join(
+        os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
+        "data",
+        "ukb_ref.csv",
+    )
+    df_ref = pd.read_csv(path_to_ref_csv, sep=";")
+
+    df_clinical = df_clinical.set_index(
+        ["participant_id", "session", "suffix"], verify_integrity=True
+    )
+
+    # Build participants dataframe
+    df_participants = df_clinical.filter(items=list(df_ref["participants"]))
+
+    # Build sessions dataframe
+    df_session = df_clinical.filter(items=list(df_ref["session"]))
+
+    # Build scans dataframe
+    df_scan = df_clinical.filter(items=list(df_ref["scan"]))
+    return df_participants, df_session, df_scan
+
+
+def write_to_tsv(dataframe: DataFrame, buffer: Union[PathLike, BinaryIO]) -> None:
+    # Save dataframe as a BIDS-compliant TSV file.
+    dataframe.to_csv(buffer, sep="\t", na_rep="n/a", date_format="%Y-%m-%d")
+
+
+def write_bids(
+    to: PathLike,
+    participants: DataFrame,
+    sessions: DataFrame,
+    scans: DataFrame,
+    dataset_directory: PathLike,
+) -> List[PathLike]:
+    from pathlib import Path
+
+    from fsspec.implementations.local import LocalFileSystem
+
+    from clinica.iotools.bids_dataset_description import BIDSDatasetDescription
+
+    to = Path(to)
+    fs = LocalFileSystem(auto_mkdir=True)
+
+    participants = participants.droplevel(["session", "suffix"]).drop_duplicates()
+
+    # Ensure BIDS hierarchy is written first.
+    with fs.transaction:
+        with fs.open(to / "dataset_description.json", "w") as dataset_description_file:
+            BIDSDatasetDescription(name="UKB").write(to=dataset_description_file)
+
+        with fs.open(to / "participants.tsv", "w") as participant_file:
+            write_to_tsv(participants, participant_file)
+
+    for participant_id, data_frame in sessions.groupby(["participant_id"]):
+        session = data_frame.droplevel(["participant_id", "suffix"]).drop_duplicates()
+
+        session_filepath = to / participant_id / f"{participant_id}_sessions.tsv"
+        with fs.open(session_filepath, "w") as sessions_file:
+            write_to_tsv(session, sessions_file)
+
+    scans = scans.set_index(["bids_full_path"], verify_integrity=True)
+    for bids_full_path, metadata in scans.iterrows():
+        copy_file_to_bids(
```

Comment:
Looks like this can be factorized easily

Reply:
done !